ROC curves for multi-class (What-If Tool)

grovina opened the pull request:
* Motivation for features / changes

Show ROC curves for multi-class models.


* Technical description of changes

For each class, we can plot a ROC curve considering the class in question as the positive one and all the others as negatives (what is called a *binarized* version of the problem).
To achieve this, we iterate through all examples, populating an object with the following structure:
```> model > slice > label > threshold > classification stats```
Where the classification data contain the number and rates of true and false positives and negatives. This is done for every value of the threshold (every 1% between 0 to 100%), for every class label, for each feature slice (either each value for categorical features or each interval for numeric features), for every model.
The overall case (not sliced by any feature) is trivially treated as a single slice.

When displaying the ROC curves, we end up with the same case as the ordinary ROC curves from binary classification problems, once for each slice.


* Screenshots of UI changes

The overall case:
![Screenshot from 2019-10-10 14-34-36](https://user-images.githubusercontent.com/6004563/66573773-1a7df080-eb6b-11e9-89f3-890fa26a741d.png)

The sliced feature case:
![Screenshot from 2019-10-10 14-34-08](https://user-images.githubusercontent.com/6004563/66573774-1a7df080-eb6b-11e9-9931-28898e489e38.png)


* Detailed steps to verify changes work correctly (as executed by you)
Open performance tab in the iris demo, and check that:
   - the correct number of ROC curves appear (same number as number of classes in the problem)
   - the values and shape of the ROC curve are coherent
   - the curve respects the data

Repeat for sliced features, and also for different numbers of buckets

* Alternate designs / implementations considered

We could think of averaged ROC curves (like https://scikit-learn.org/stable/auto_examples/model_selection/plot_roc.html). I personally think they are a bit harder to interpret.

Another possibility is merging all curves into a single plot. Although more compact, it wouldn't be as clear that each curve refers to a distinct version of the problem, and it could also get a bit too crowded with many classes or multiple models.

grovina commented:
cc @jameswex and @tolga-b

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -4913,6 +5074,53 @@ <h2>Show similarity to selected datapoint</h2>
+                  plotThresholds,
+                  regenInferenceStats,
+                  true
+                );
```
should really do ROC and PR curves for each class just like with binary

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -4854,7 +5015,7 @@ <h2>Show similarity to selected datapoint</h2>
           this.featureValueThresholds = this.sortFeatureValues(tempArray);
 
-          // ROC curves should only exist for the binary case
+          // ROC curves for the binary case
```
update comment since also does PR curves

jameswex commented on tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html:
```diff
@@ -5864,6 +6072,14 @@ <h2>Show similarity to selected datapoint</h2>
+          return this.getRocChartId(index) + '-' + label;
+        },
+
+        getLabelVocab: function(index) {
```
call this getLabel or getLabelForIndex for clarity

jameswex commented:
thanks @grovina , just the one comment about padding now

jameswex commented:
also make sure to run lint

grovina commented:
Adjusted some CSS and fixed lint. Here screenshots for:

* the multi-class case:
![Screenshot from 2019-10-11 17-38-12](https://user-images.githubusercontent.com/6004563/66668785-ea5b4e00-ec4d-11e9-8432-4f75af09d9d3.png)

* the binary case:
![Screenshot from 2019-10-11 17-37-53](https://user-images.githubusercontent.com/6004563/66668788-ed563e80-ec4d-11e9-8fa1-5ab56de18f1a.png)
